File: app/lib/tools/validator/email.dart
/*
** Copyright GETOUT SAS - All Rights Reserved
** Unauthorized copying of this file, via any medium is strictly prohibited
** Proprietary and confidential
** Wrote by Perry Chouteau <perry.chouteau@epitech.eu>
*/

import 'package:flutter/cupertino.dart';
import 'package:getout/constants/regular_expression.dart';
import 'package:getout/tools/app_l10n.dart';

String? emailValidator(BuildContext context, String? email) {
  RegExp regex = RegExp(RegularExpression.email);

  if (email == null || !regex.hasMatch(email)) {
    return appL10n(context)?.email_validator;
  }
  return null;
}

String? confirmEmailValidator(BuildContext context, String? newEmail, String? confirmEmail) {
  String? result = emailValidator(context, confirmEmail);

  if (result != null) {
    return result;
  }
  return (newEmail != confirmEmail) ? appL10n(context)?.email_matching : null;
}

File: backend/src/models/account/email.ts
/*
** Copyright GETOUT SAS - All Rights Reserved
** Unauthorized copying of this file, via any medium is strictly prohibited
** Proprietary and confidential
** Wrote by Alexandre Chetrit <chetrit.pro@hotmail.com>
*/

import { type UUID } from 'crypto'
import { StatusCodes } from 'http-status-codes'

import { sendEmailVerificationEmail } from '@services/brevo/emails'
import logger from '@services/middlewares/logging'
import { AccountDoesNotExistError, DbError } from '@services/utils/customErrors'

import { findEntity } from '@models/getObjects'

import { Account } from '@entities/Account'

import { appDataSource } from '@config/dataSource'

import { generateEmailVerificationCode } from './verifyEmail'

async function changeAccountEmail (accountId: UUID, newEmail: string): Promise<void> {
  await findEntity<Account>(Account, { id: accountId }).then(async (account: Account | null) => {
    if (account == null) {
      throw new AccountDoesNotExistError(undefined, StatusCodes.NOT_FOUND)
    }
    account.newEmail = newEmail
    return await appDataSource.getRepository<Account>('Account').save(account)
  }).then(async (account: Account | null) => {
    if (account == null) {
      throw new DbError('Failed modifying the new email.')
    }
    return await generateEmailVerificationCode(account.email).then(async (code) => {
      logger.debug(`Sending email verification code to new email ${account.newEmail} : ${code}.`)
      return await sendEmailVerificationEmail(account, code, true)
    })
  })
}

async function verifyNewEmail (accountId: UUID): Promise<void> {
  await findEntity<Account>(Account, { id: accountId }).then(async (account: Account | null) => {
    if (account == null) {
      throw new AccountDoesNotExistError(undefined, StatusCodes.NOT_FOUND)
    }
    account.email = account.newEmail!
    account.newEmail = null
    return await appDataSource.getRepository<Account>('Account').save(account)
  })
}

export { changeAccountEmail, verifyNewEmail }
